File: src/device.py
# Copyright (C) Hyogi Sim <hyogi@cs.vt.edu>
# 
# ---------------------------------------------------------------------------
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
# 

import array

class StorageDevice(object):
    def __init__(self, capacity, block_size, read_latency, write_latency):
        self.__capacity = capacity
        self.__block_size = block_size
        self.__read_latency = read_latency
        self.__write_latency = write_latency
        self.__read_count = 0
        self.__write_count = 0

    @property
    def read_latency(self):
        return self.__read_latency

    @property
    def write_latency(self):
        return self.__write_latency

    @property
    def capacity(self):
        return self.__capacity

    @property
    def block_size(self):
        return self.__block_size

    @property
    def read_count(self):
        return self.__read_count

    @property
    def write_count(self):
        return self.__write_count

    def read_block(self, block, count):
        self.__read_count += 1
        return self.read_latency;

    def write_block(self, block, count):
        self.__write_count += 1
        return self.write_latency;

class Memory(StorageDevice):
    def __init__(self, capacity, block_size, \
                read_latency=0, write_latency=0):
        StorageDevice.__init__(self, capacity, block_size, read_latency, \
                            write_latency)

class SSD(StorageDevice):
    """ In SSD class, we keep track of write distribution among blocks.
    """
    def __init__(self, capacity, block_size, read_latency, write_latency):
        StorageDevice.__init__(self, capacity, block_size, read_latency, \
                            write_latency)

        # if this overflows, we need to use c_int64
        self.block_wear = array.array('L', map(lambda x:0, range(capacity)))

    def write_block(self, block, count):
        self.__write_count += 1
        for i in range(block, block + count):
            self.block_wear[i] += 1
        return self.__write_latency

class HDD(StorageDevice):
    def __init__(self, capacity, block_size, \
                read_latency=100.0, write_latency=100.0):
        StorageDevice.__init__(self, capacity, block_size, read_latency, \
                            write_latency)

# Test driver
if __name__ == '__main__':
    pass

File: src/device.c
/* Copyright (C) Hyogi Sim <hyogi@cs.vt.edu>
 * 
 * ---------------------------------------------------------------------------
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * 
 */
#include <stdlib.h>
#include "device.h"

struct storage *storage_init(__u32 node,
			__u64 block_size, __u64 block_count,
			__u32 latency_read, __u32 latency_write,
			int wear,
			struct storage_operations *ops)
{
	struct storage *self = NULL;
	size_t len = sizeof(*self);

	if (wear)
		len += sizeof(__u64) * block_count;

	self = malloc(len);
	if (self) {
		self->node = node;
		self->block_size = block_size;
		self->block_count = block_count;
		self->capacity = block_size * block_count;
		self->latency_read = latency_read;
		self->latency_write = latency_write;
		self->stat_reads = self->stat_writes = 0;
		self->ops = ops;

		if (!wear)
			self->wear = NULL;
	}

	return self;
}


void storage_exit(struct storage *self)
{
	if (self)
		free(self);
}

static __u64 generic_read_block(struct storage *self, __u64 offset, __u64 len)
{
	self->stat_reads += len;

	return self->latency_read * len;
}

static __u64 generic_write_block(struct storage *self, __u64 offset, __u64 len)
{
	self->stat_writes += len;

	return self->latency_write * len;
}

struct storage_operations generic_storage_ops = {
	.read_block	= &generic_read_block,
	.write_block	= &generic_write_block,
};

